internal/iamauth/auth.go

```diff
@@ -0,0 +1,336 @@
+		config: config,
+		logger: logger,
+	}, nil
+}
+
+// ValidateLogin determines if the identity in the loginToken is permitted to login.
+// If so, it returns details about the identity. Otherwise, an error is returned.
+func (a *Authenticator) ValidateLogin(ctx context.Context, loginToken string) (*IdentityDetails, error) {
+	token, err := NewBearerToken(loginToken, a.config)
+	if err != nil {
+		return nil, err
+	}
+
+	req, err := token.GetCallerIdentityRequest()
+	if err != nil {
+		return nil, err
+	}
+
+	if a.config.ServerIDHeaderValue != "" {
+		err := validateHeaderValue(req.Header, a.config.ServerIDHeaderName, a.config.ServerIDHeaderValue)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	callerIdentity, err := a.submitCallerIdentityRequest(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+
+	entity, err := responses.ParseArn(callerIdentity.Arn)
+	if err != nil {
+		return nil, err
+	}
+
+	identityDetails := &IdentityDetails{
+		EntityName: entity.FriendlyName,
+		EntityId:   callerIdentity.UserId,
+		AccountId:  callerIdentity.Account,
+	}
+	clientArn := entity.CanonicalArn()
+
+	// Fetch the IAM Role or IAM User, if configured.
+	// This requires the token to contain a signed iam:GetRole or iam:GetUser request.
+	if a.config.EnableIAMEntityDetails {
+		iamReq, err := token.GetEntityRequest()
+		if err != nil {
+			return nil, err
+		}
+
+		if a.config.ServerIDHeaderValue != "" {
+			err := validateHeaderValue(iamReq.Header, a.config.ServerIDHeaderName, a.config.ServerIDHeaderValue)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		iamEntityDetails, err := a.submitGetIAMEntityRequest(ctx, iamReq, token.iamEntityType)
+		if err != nil {
+			return nil, err
+		}
+
+		// Only the CallerIdentity response is a guarantee of the client's identity.
+		// The role/user details must have a unique id match to the CallerIdentity before use.
+		if iamEntityDetails.EntityId() != callerIdentity.UserId {
+			return nil, fmt.Errorf("unique id mismatch in login token")
+		}
+
+		// Use the full ARN with path from the Role/User details
+		clientArn = iamEntityDetails.EntityArn()
+		identityDetails.EntityPath = iamEntityDetails.EntityPath()
+		identityDetails.EntityTags = iamEntityDetails.EntityTags()
+	}
+
+	a.logger.Info("iamauth login attempt", "arn", clientArn)
+	if err := a.validateIdentity(clientArn); err != nil {
+		return nil, err
+	}
+	return identityDetails, nil
+}
+
+// https://github.com/hashicorp/vault/blob/ba533d006f2244103648785ebfe8a9a9763d2b6e/builtin/credential/aws/path_login.go#L1321-L1361
+func (a *Authenticator) validateIdentity(clientArn string) error {
+	if strutil.StrListContains(a.config.BoundIAMPrincipalARNs, clientArn) {
+		// Matches one of BoundIAMPrincipalARNs, so it is trusted
+		return nil
+	}
+	if a.config.EnableIAMEntityDetails {
+		for _, principalArn := range a.config.BoundIAMPrincipalARNs {
+			if strings.HasSuffix(principalArn, "*") && strutil.GlobbedStringsMatch(principalArn, clientArn) {
```

Comment:
Should be pretty easy to lift `GlobbedStringsMatch` over into `github.com/hashicorp/consul/lib` to avoid needing a little dependency for just two functions (one of which already exists in `lib/stringslice`)